opentelemetry-api/src/opentelemetry/metrics/__init__.py

```diff
@@ -0,0 +1,343 @@
+# limitations under the License.
+
+# pylint: disable=too-many-ancestors
+# type: ignore
+
+# FIXME enhance the documentation of this module
+"""
+This module provides abstract and concrete (but noop) classes that can be used
+to generate metrics.
+"""
+
+
+from abc import ABC, abstractmethod
+from logging import getLogger
+from os import environ
+from threading import RLock
+from typing import Optional, cast
+
+from opentelemetry.environment_variables import OTEL_PYTHON_METER_PROVIDER
+from opentelemetry.metrics.instrument import (
+    Counter,
+    DefaultCounter,
+    DefaultHistogram,
+    DefaultObservableCounter,
+    DefaultObservableGauge,
+    DefaultObservableUpDownCounter,
+    DefaultUpDownCounter,
+    Histogram,
+    ObservableCounter,
+    ObservableGauge,
+    ObservableUpDownCounter,
+    UpDownCounter,
+)
+from opentelemetry.util._providers import _load_provider
+
+_logger = getLogger(__name__)
+
+
+class MeterProvider(ABC):
+    def __init__(self):
+        super().__init__()
+        self._lock = RLock()
+
+    @abstractmethod
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> "Meter":
+        if name is None or name == "":
+            _logger.warning("Invalid name: %s", name)
+
+
+class _DefaultMeterProvider(MeterProvider):
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> "Meter":
+        with self._lock:
+            super().get_meter(name, version=version, schema_url=schema_url)
+            # This is done in order to make it possible to store configuration
+            # in the meter provider and make it automatically accessible for
+            # any meter even after it changes.
+            meter = _DefaultMeter(name, version=version, schema_url=schema_url)
+            meter._meter_provider = self  # pylint: disable=protected-access
+            return meter
+
+
+class ProxyMeterProvider(MeterProvider):
+    def get_meter(
+        self,
+        name,
+        version=None,
+        schema_url=None,
+    ) -> "Meter":
+        with self._lock:
+            if _METER_PROVIDER:
+                return _METER_PROVIDER.get_meter(
+                    name, version=version, schema_url=schema_url
+                )
+            return ProxyMeter(name, version=version, schema_url=schema_url)
+
+
+class Meter(ABC):
+    def __init__(self, name, version=None, schema_url=None):
+        super().__init__()
+        self._lock = RLock()
+        self._name = name
+        self._version = version
+        self._schema_url = schema_url
+        self._instrument_names = set()
+        self._meter_provider = None
+
+    @property
+    def name(self):
+        with self._lock:
+            return self._name
+
+    @property
+    def version(self):
+        with self._lock:
+            return self._version
+
+    @property
+    def schema_url(self):
+        with self._lock:
+            return self._schema_url
```

Comment:
We need these attributes being defined as properties in the API because of this specific [requirement](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/api.md#get-a-meter):

_In case an invalid name (null or empty string) is specified, a working Meter implementation MUST be returned as a fallback rather than returning null or throwing an exception, **its name property SHOULD keep the original invalid value**, and a message reporting that the specified value is invalid SHOULD be logged. _

So, in order for us to make sure that the original value has been kept, it must be an interface for the `name` property.

Reply:
This doesn't imply you need to expose it as a publicly-gettable concern.  The name only needs to be kept for SDK usage.

Reply:
We need to expose it as a public method. The SDK subclasses would be able to use it only if it is public, Python lacks some intermediate protection access equivalent to `protected` in Java.

Reply:
[This](https://github.com/open-telemetry/opentelemetry-python/pull/1887#discussion_r710248608) conversation made me look deeper into what is considered to be a public, protected and private attribute in Python. So far I have been under the impression that there is no official way to define a protected attribute but apparently there is at least a convention on how to do it:

-. `public_attribute`
-. `_protected_attribute`
-. `__private_attribute`

These are some places where I have found this convention:

https://towardsdatascience.com/private-protected-attributes-in-python-demystified-once-and-for-all-9456d4e56414
https://www.tutorialsteacher.com/python/public-private-protected-modifiers
https://www.geeksforgeeks.org/protected-variable-in-python/

None of them is part of the official documentation. The closest I have found is
this:

https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables

I wish that part of the PEP was written with stronger specificity, it does not
explicitly say that protected attributes are to be prefixed by a single
underscore but it looks like it can be deduced from its wording.

The word "protected" is only mentioned once in that PEP, wich uses the term
"subclass API" to refer to the concept of protected in other languages. It
then goes to say that attributes that should not be part of the subclass API
should be named with two leading underscores.

Anyways, I think we have found at least one use case where using this
convention would be convenient. At least I would like to have a way to separate
from public, protected and private so I suggest we follow that convention in
our project.

Reply:
> The SDK subclasses would be able to use it only if it is public

Aren't we accessing the protected attrs in SDK and Exporters already?

Reply:
> > The SDK subclasses would be able to use it only if it is public
> 
> Aren't we accessing the protected attrs in SDK and Exporters already?

Yes, I have explained my new point of view in the comment above, I am ok with following the convention of a single leading underscore for protected attributes.